### ISSUE TITLE
[hue] "scene" channel on a hue group populated with several minutes delay

### ISSUE THREAD

**Issue by mhueschen** [openhab/openhab-addons#14137]
Happy new year everyone.

The hue channel type ID "scene" can be used on the bridge or on a hue group. On the bridge it lists all scenes on the bridge with room names in brackets and when used on a hue group it filters the scenes to the scenes in this group. This worked as expected until Milestone M3.

Now it only works at the bridge level. The scene list for a hue group remains empty. Unfortunately this is the most useful feature as you want to select a scene for a room from a filtered list for this room (hue group).

I now downgraded to M3 as this is an essential feature more me. Hopefully there is someone with code knowledge of the hue binding who can support here.

Thanks in advance.

**Comment by lolodomo** [comment]
Can you check these console commands (using your own thing UIDs):
```
hue hue:bridge:bridge1 scenes
hue hue:group:bridge1:salon scenes
```
Both are returning a list of scenes in my case.
And commands for the scene channel on my hue group thing are present in MainUI and BasicUI.
But I am using a bridge V1, maybe a firmware update of bridge V2 broke something ?
I have a serious doubt that anything was changed in the hue binding regarding scenes between 3.4M3 and final 3.4.

**Comment by mhueschen** [comment]
Thanks for the quick reply.
I ran both commands using my own bridge UID and a sample group UID.

3.4M3 gives me correct output for bridge and my group.
3.4 Final gives correct output for the bridge and no output for the group.

So the behaviour can be reproduced on the console. at least in my setting.

**Comment by lolodomo** [comment]
> I suppose a firmware update change of the hue bridge would have given the same results on both builds.

Good remark. 

So I have no other idea. 
To be analyzed by someone who can reproduce your problem. 
Maybe @cweitkamp or @jlaur ?

**Comment by mhueschen** [comment]
Just to add a bit more context I did a quick check on build 3.4M4 and it has the same problem.

So I am pretty sure that the relevant code change that introduced the bug happened between 3.4M3 and M4.

**Comment by lolodomo** [comment]
The culpit would be very probably PR #11842 which was merged in milestone 4.

**Comment by mhueschen** [comment]
I tried to get a bit more light into this myself. I apologize upfront for maybe going in a wrong direction but I think I have found a possible cause. I am a .NET developer for many years and really not used to JAVA and Eclipse and everything involved with that. But I managed to get a dev env working and debugging the current Hue Binding source code.

What I have found seems to be a timing error in the HueBridgeHandler.
Line 292 "updateGroups" gets run after(!) Line 1070 in the "notifyGroupSceneUpdate". Therefore the groupStatusListeners are still empty and bot build for the groups, so they cannot get the retrieved scenes.

Does that make sense to anyone.

**Comment by mhueschen** [comment]
UPDATE:
I think my thoughts are correct. I found in the source code that the scenes are updated every 10 minutes. So I installed the new build and run it..... no scenes for my groups. But aft6er waiting 10 minutes they are there. Honestly I never waited 10 minutes when I checked the newer builds before.

So...the feature is working as expected. BUT only after 10 minutes after an openhab restart. Because the first scene 
refresh is done to early before the groups are ready to receive the updated scenes.

I am far from really understanding the code or the timing that happens. So maybe one of the usual maintainer of the binding can fix this now with these hints.

**Comment by lolodomo** [comment]
Your analysis is perfectly correct.
I do not remember if I had that possible delay in mind when implementing the groups and scenes, or if it evolves. That is not ideal I agree.

We should at least mention it in the documentation. 

And as the binding is more or less rewritten in OH 4, we can hope a better implementation of this case.

**Comment by mhueschen** [comment]
That makes sense. It is a bit annoying to wait 10 minutes after each openhab restart until this works (as the scene refresh interval is not a setting), but it's fine for me until the binding is rewritten anyway in OH4. Really looking forward to this and the support for all features of the newer Hue API as I have 4 hue bridges and way over 100 hue devices. Appreciation for you @lolodomo here for implementing the scenes feature to the binding. That made my life way better :-)

So maybe you could make a commit and pull request for the docs update? I am not yet used to this here. I think then we could close this issue.

Great to see that it helped to find the issue by digging into the code.

**Comment by lolodomo** [comment]
The initial PR: https://github.com/openhab/openhab-addons/pull/8098

Apparently the risk was assumed, look at this comment:
https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/HueBridgeHandler.java#L662
There is a delay of 5 seconds, probably not sufficient in certain circumstances. We could increase it to reduce the risk.

**Comment by mhueschen** [comment]
Oh right.
Maybe this could also be a setting for the Bridge Thing. So everyone could adjust it to their needs.

**Comment by jlaur** [comment]
> And as the binding is more or less rewritten in OH 4, we can hope a better implementation of this case.

The new API v2 implementation will only apply to Hue Bridge 2.0+ though. Perhaps the loading should be chained, so instead of using delays, we would load things in the right order? I should add that I haven't looked into the details.

**Comment by lolodomo** [comment]
> Perhaps the loading should be chained, so instead of using delays, we would load things in the right order? I should add that I haven't looked into the details.

No. In fact, getting the list of scenes is something required for the bridge thing and for each hue group thing.

A dirty and easy solution would be to delay enough the request run by the bridge to be sure that all hue group things are initialized. In practice, 30 seconds (instead of 5) would be certainly enough.

The proper solution would be to let the bridge run its request with no delay and then cache the result. When a hue group thing is initialized and as a consequence registered to the bridge thing as listener for scenes, the cached value is used (if defined).

Note that with the current solution and the dirty patch solution, if you add a new hue group thing later after the OH startup, you will have until 10 minutes to wait to get the list of scenes for this thing.

I can try to implement the proper solution.

**Comment by jlaur** [comment]
> What is important is that it is absolutely not a critical issue, the current solution is just to wait 10 minutes.
> And this is not a bug that was introduced in 3.4. Could have occurred even in previous versions.

Agreed.

**Comment by lolodomo** [comment]
I updated the issue title to reflect the conclusion of the analysis.

**Comment by mhueschen** [comment]
Thanks to all of you.

Especially seeing that my most important binding is maintained and developed in such an active manner. Looking forward to the next steps.

**Comment by lolodomo** [comment]
The problem is not exactly what I initially thought and it was not easy to understand (I added a lot of logs).
The problem is that the first time scenes are retrieved, groups are not yet retrieved. This is due to the fact that the first try to get lights/groups, it fails because the binding is not yet authenticated.
This is clear for me now, needs to find the best fix.

**Comment by lolodomo** [comment]
I think we do not need a specific scheduled thread for scenes, I will include that in the lights/groups scheduled thread and avoid to request the list of scenes to much often.

**Comment by lolodomo** [comment]
> Especially seeing that my most important binding is maintained and developed in such an active manner. Looking forward to the next steps.

@mhueschen : for your information, the enhancement is available in snapshot 3301 (OH 4.0).
